benchcab/spatial.py

```diff
@@ -0,0 +1,169 @@
+
+
+class SpatialTask:
+    """A class used to represent a single spatial task."""
+
+    root_dir: Path = internal.CWD
+    subprocess_handler: SubprocessWrapperInterface = SubprocessWrapper()
+    git_repo_handler: GitRepoInterface = GitRepo()
+
+    def __init__(
+        self,
+        repo: CableRepository,
+        met_forcing_name: str,
+        met_forcing_payu_experiment: str,
+        sci_conf_id: int,
+        sci_config: dict,
+        payu_args: Optional[str] = None,
+    ) -> None:
+        self.repo = repo
+        self.met_forcing_name = met_forcing_name
+        self.met_forcing_payu_experiment = met_forcing_payu_experiment
+        self.sci_conf_id = sci_conf_id
+        self.sci_config = sci_config
+        self.payu_args = payu_args
+
+    def get_task_name(self) -> str:
+        """Returns the file name convention used for this task."""
+        return f"{self.met_forcing_name}_R{self.repo.repo_id}_S{self.sci_conf_id}"
+
+    def setup_task(self, payu_config: dict = None, verbose=False):
+        """Does all file manipulations to run cable with payu for this task."""
+
+        if verbose:
+            print(f"Setting up task: {self.get_task_name()}")
+
+        task_dir = self.root_dir / internal.SPATIAL_TASKS_DIR / self.get_task_name()
+
+        self.git_repo_handler.clone_from(
+            self.met_forcing_payu_experiment,
+            task_dir.relative_to(self.root_dir),
+            verbose=verbose,
+        )
+
+        exp_config_path = task_dir / "config.yaml"
+        with exp_config_path.open("r", encoding="utf-8") as file:
+            config = yaml.safe_load(file)
+            if config is None:
+                config = {}
+
+        if verbose:
+            print(
+                "  Updating experiment config parameters in",
+                internal.SPATIAL_TASKS_DIR / self.get_task_name() / "config.yaml",
+            )
+
+        if payu_config:
+            config = deep_update(config, payu_config)
+
+        config["exe"] = str(
+            self.root_dir
+            / internal.SRC_DIR
+            / self.repo.name
+            / "offline"
+            / internal.CABLE_MPI_EXE
+        )
+
+        # Here we prepend inputs to the `input` list so that payu knows to use
+        # our inputs over the pre-existing inputs in the config file:
+        config["input"] = [
+            str(
+                self.root_dir
+                / internal.CABLE_AUX_DIR
+                / "core"
+                / "biogeophys"
+                / "def_veg_params_zr_clitt_albedo_fix.txt"
+            ),
+            str(
+                self.root_dir
+                / internal.CABLE_AUX_DIR
+                / "core"
+                / "biogeophys"
+                / "def_soil_params.txt"
+            ),
+            *config.get("input", []),
+        ]
+
+        config["laboratory"] = str(self.root_dir / internal.PAYU_LABORATORY_DIR)
```

Comment:
> Payu is built with the design to have one laboratory for every run of a specific model (if I get this right). So instead of creating a laboratory specific to benchcab (i.e. within the benchcab work directory), why not use the default payu laboratory at: /scratch/$PROJECT/$USER/cable ?

The use case I had in mind was if we want to reset the `bench_example` directory, we can `rm -rf` the `runs` directory without having to worry about any side effects that might have occurred outside of the `bench_example` directory.

> Also, in the documentation is says to setup the laboratory first and I don't seem to see that in benchcab. I haven't run payu enough to know if it is really an issue or if it's handled differently.

Thanks for pointing this out. I will see if the laboratory needs to be set up manually prior to running benchcab.